scripts/polyrith_sage.py

```diff
@@ -0,0 +1,136 @@
+        else:
+            out.append("(poly.mul " + const_string(coeffs[i]) + " " + monomial_to_string(monomials[i]) + ")")
+    return poly_terms_string(out)
+'''
+
+# future extensions may change behavior depending on the base type
+def type_str(type):
+    return "QQ"
+
+def var_names(var_list_string):
+    return var_list_string[1:-1].replace(" ", "")
+
+def create_query(type: str, var_list, eq_list, goal_type):
+    """ Create a query to invoke Sage's `MPolynomial_libsingular.lift`. See
+    https://github.com/sagemath/sage/blob/f8df80820dc7321dc9b18c9644c3b8315999670b/src/sage/rings/polynomial/multi_polynomial_libsingular.pyx#L4472-L4518
+    for a description of this method. """
+    query = f'''
+{var_names(var_list)} = {type_str(type)}['{var_names(var_list)}'].gens()
+gens = {eq_list}
+p = {goal_type}
+I = ideal(gens)
+coeffs = p.lift(I)
+print(list(map(polynomial_to_string, coeffs)))
+'''
+    return query
+
+class EvaluationError(Exception):
+    def __init__(self, ename, evalue, message='Error in Sage communication'):
+        self.ename = ename
+        self.evalue = evalue
+        self.message = message
+        super().__init__(self.message)
+
+def evaluate_in_sage(query: str, format=False) -> str:
+    if format:
+        clean_query = query
+        query = (f'print({clean_query})')
+    data = {'code': query}
+    headers = {'content-type': 'application/x-www-form-urlencoded'}
+    response = requests.post('https://sagecell.sagemath.org/service', data, headers=headers)
+    response = json.loads(response.text)
+    if response['success']:
+        return response['stdout'] if 'stdout' in response else None
+    elif 'execute_reply' in response and 'ename' in response['execute_reply'] and 'evalue' in response['execute_reply']:
+        raise EvaluationError(response['execute_reply']['ename'], response['execute_reply']['evalue'])
+    else:
+        raise Exception(response)
+
+def main():
+    '''The system args contain the following:
+    0 - the path to this python file
+    1 - a string containing "tt" or "ff" depending on whether polyrith was called with trace enabled
+    2 - a string representing the base type of the target
+    3 - a list of all the variables to be used
+    4 - a list of the polynomial hypotheses/proof terms in terms of the variables
+    5 - a single polynomial representing the target'''
+    command = create_query(sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5])
+    final_query = polynomial_formatting_functions + "\n" + command
+    if sys.argv[1] == 'tt': # trace enabled
+        print(command)
+    else:
+        try:
+            output = evaluate_in_sage(final_query).replace("'", "")
+        except EvaluationError as e:
+            print(f'%{e.ename}%{e.evalue}')
+        else:
+            output = output.replace(",", "")
+            output = output.replace("[", "").replace("]", "").strip()
+            output += " "
+            print(output)
```

Comment:
This post-processing looks awfully strange to me. Why not just

```suggestion
        try:
            output = dict(success=True, data=evaluate_in_sage(final_query).replace("'", ""))
        except EvaluationError as e:
            output = dict(success=False, error_name=e.ename, error_value=e.value)
        print(json.dumps(output))
```
and then use the lean json parser at the other end

Reply:
@robertylewis, I can push through this change if you'd like

Reply:
JSON communcation might also simplify the tracing changes, but I suspect it isn't a small change.